monitoring/space_monitor.py

```diff
@@ -0,0 +1,251 @@
+    t_reminder = reminder_days * (60 * 60 * 24)  # in seconds
+
+    spaces = datasets + users
+
+    last_status = load_last_status()
+    for space in set(spaces) - set(last_status.keys()):
+        logging.warn('Space "%s" was not in last status.' % space)
+        last_status[space] = 'ok'
+
+    unknown_spaces = set(spaces) - set(get_status(datasets, users).keys())
+    if unknown_spaces:
+        logging.error('Unknown spaces: %s' % ' '.join(unknown_spaces))
+        sys.exit('Unrecoverable error. Please check the log file.')
+
+    last_notification_time = {space: 0 for space in spaces}
+
+    while True:
+        logging.debug('Checking space status...')
+        status = get_status(datasets, users)
+        for space in spaces:
+            status_new, status_old = status[space], last_status[space]
+            dt = time.time() - last_notification_time[space]
+            if alert(space, status_new, status_old, remind=(dt > t_reminder)):
+                notify(space, status_new, status_old)
+                last_notification_time[space] = time.time()
+        last_status = status
+        save_status(status)
+        time.sleep(interval)
+
+
+def load_last_status():
+    "Return status as loaded from disk"
+    logging.debug('Getting initial status...')
+    try:
+        status = {}
+        for line in open(STATUSFILE):
+            s, m = line.strip().split(' ', 1)
+            status[m] = s
+        return status
+    except Exception as e:  # anyone, really, just be robust
+        logging.info('Could not read last status: %s' % e)
+        return {}
+
+
+def save_status(status):
+    "Save the given status to disk"
+    with open(STATUSFILE, 'wt') as f:
+        for m, s in status.items():
+            f.write('%s %s\n' % (s, m))
+
+
+def alert(space, status_new, status_old, remind):
+    "Return True if the status of the space requires a notification"
+    if status_new == status_old != 'warn':
+        logging.debug('Nothing to report on space "%s".' % space)
+        return False
+    elif status_new != status_old:
+        logging.warn('Changes found in space "%s". Notifying.' % space)
+        return True
+    elif status_new == 'warn' and remind:
+        logging.warn('Error persists in space "%s". Notifying.' % space)
+        return True
+    else:
+        logging.debug('Status of space "%s": %s' % (space, status_new))
+        return False
+
+
+def get_status(datasets, users):
+    "Return dict with dataset/user and its status (ok, warn, undefined)"
+    # ok if less than 80% of space used, warn if more.
+    status = {}
+    for dataset, ratio in get_ratio_datasets().items():
+        if dataset in datasets:
+            status[dataset] = 'ok' if ratio < TRIGGER_FRACTION else 'warn'
+    for user in users:
+        try:
+            status[user] = ('ok' if get_ratio_user(user) < TRIGGER_FRACTION
+                            else 'warn')
+        except ZeroDivisionError:
+            status[user] = 'undefined'
```

Comment:
The first time you run the script (no last_space_status.txt exists), if a user has no quota, the status result is `undefined` and the user gets a notification. I guess the user should not be notified in this case. On the next runs, there is no problem, the status file has "undefined USER" and since there is no change of state, no email is sent.